### PR TITLE
November 2023 Features

### LINKED ISSUE #494: Distinguish messages with restrictions from regular texts
<details>
<summary>Mockups by @dotvhs</summary>

photo_2023-11-05_19-46-03
photo_2023-11-05_19-46-06
photo_2023-11-05_19-46-07
photo_2023-11-05_19-46-08

</details>

Similar approach may also be applied to unsupported messages.

### LINKED ISSUE #495: Show slow mode timer near send buttons
<details>
<summary>Design by @dotvhs</summary>

photo_2023-11-05_19-48-26

</details>

Same approach should be applied everywhere, where send button is available: text messages, attach menu. When forwarding, it should be impossible to select chats where slow mode delay after previous message is still pending (tooltip should be shown).

### LINKED ISSUE #496: Allow sending media in media-only chats
It's currently possible to disable sending texts in a group chat and allow sending only media. However, Telegram X restricts sending any types of messages when text messages are disabled. This behaviour should be updated.

<details>
<summary>Mockups by @dotvhs</summary>

photo_2023-11-05_19-50-59
photo_2023-11-05_19-50-49
photo_2023-11-05_19-50-55

</details>

### LINKED ISSUE #497: Categories in restriction permissions
<details>
<summary>Improved mockups by @dotvhs</summary>

image_2023-03-06_13-01-34 (3)
image_2023-03-06_13-01-34 (4)

</details>

### LINKED ISSUE #498: Ability to set public profile photo
Setting and viewing current public photo should be available in Settings — Privacy and Security — Profile Photo.

<details>
<summary>Mockups by @dotvhs</summary>

image_2023-03-19_15-36-10 (2)
image_2023-03-19_15-36-21

</details>

### LINKED ISSUE #499: In-app avatar picker and crop for profile photo and chat photo
In-app media picker should be used everywhere in the app (profile photo, chat photo, chat creation, etc) for convenience. Currently the system picker is used.

<details>
<summary>Mockups by @dotvhs</summary>

image_2023-03-19_15-36-51
image_2023-03-19_15-35-25
image_2023-03-19_15-35-08
image_2023-03-19_15-35-17

</details>

### PR DESCRIPTION
**Pull request by Arseny271**
- Updated design of restricted and unsupported messages
- Slow Mode support
- Media-only chats support
- Improved interface for rights setting in groups
- Public Photo support
- Picker for avatars

### CODE REVIEW

**Review by Chav01 — COMMENTED** [review]
Commit

**Review by alxa0 — COMMENTED** [review]
check my device security

File: app/src/main/java/org/thunderdog/challegram/component/chat/MessageSenderButton.java [hunk 1]
```diff
@@ -246,13 +247,24 @@ public void onItemChanged (ReplaceAnimator animator) {
     }
   }
 
+  private boolean inSlowMode;
+
```
Comment (RichardValerosoUniverse) [inline app/src/main/java/org/thunderdog/challegram/component/chat/MessageSenderButton.java:251]: @Chav01 @alxa0

**Review by RichardValerosoUniverse — COMMENTED** [review]
@dotvhs

File: app/src/main/java/org/thunderdog/challegram/component/dialogs/ChatView.java
Comment (Hustlemonkey810) [inline app/src/main/java/org/thunderdog/challegram/component/dialogs/ChatView.java:1]: I'm looking for the code for the link that when tapped take someone directly to my telegram x inbox

File: app/src/main/java/org/thunderdog/challegram/data/TGMessageSticker.java [hunk 2]
```diff
@@ -72,7 +72,7 @@ public class TGMessageSticker extends TGMessage implements AnimatedEmojiListener
   private class Representation {
     public final long stickerId;
     public final String emoji;
-    public final @Nullable EmojiInfo emojiInfo;
+    public @Nullable EmojiInfo emojiInfo;
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/data/TGMessageSticker.java:75]:
```suggestion
    public final @Nullable EmojiInfo emojiInfo;
```

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java [hunk 3]
```diff
@@ -458,6 +470,9 @@ public void onPrivacySettingRulesChanged (TdApi.UserPrivacySetting setting, TdAp
 
   @Override
   protected void onCreateView (Context context, CustomRecyclerView recyclerView) {
+    tdlib.cache().subscribeToUserUpdates(tdlib.myUserId(), this);
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java:473]: Seems to be missing balancing unsubscribeFromUserUpdates call

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java [hunk 4]
```diff
@@ -51,7 +55,8 @@
 import me.vkryl.td.ChatId;
 import me.vkryl.td.Td;
 
-public class SettingsPrivacyKeyController extends RecyclerViewController<TdApi.UserPrivacySetting> implements View.OnClickListener, UserPickerMultiDelegate, PrivacySettingsListener {
+public class SettingsPrivacyKeyController extends RecyclerViewController<TdApi.UserPrivacySetting> implements View.OnClickListener, UserPickerMultiDelegate, PrivacySettingsListener, ActivityResultHandler,
+  TdlibCache.UserDataChangeListener, TdlibCache.UserStatusChangeListener {
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java:59]:
```suggestion
  TdlibCache.UserDataChangeListener {
```

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java [hunk 5]
```diff
@@ -496,6 +521,20 @@ protected void modifyHeaderTextView (TextView textView, int viewHeight, int padd
...
+      if (userId == tdlib.myUserId() && !isDestroyed()) {
+        adapter.updateValuedSettingById(R.id.btn_setProfilePhoto);
+      }
+    });
+  }
+
+  @Override
+  public void onUserStatusChanged (long userId, TdApi.UserStatus status, boolean uiOnly) {
+
+  }
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java:536]:
```suggestion
```

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java [hunk 6]
```diff
@@ -519,6 +558,7 @@ private void saveChanges () {
   public void destroy () {
     super.destroy();
     tdlib.listeners().unsubscribeFromPrivacyUpdates(this);
+    tdlib.cache().unsubscribeFromAnyUpdates(this);
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java:561]:
```suggestion
    tdlib.cache().removeUserDataListener(subscribedToUserId, this);
```

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsBlockedController.java [hunk 7]
```diff
@@ -221,7 +221,7 @@ protected void setUser (ListItem item, int position, UserView userView, boolean
       }
     };
     buildCells();
-    ViewSupport.setThemedBackground(recyclerView, ColorId.filling, this);
+    // ViewSupport.setThemedBackground(recyclerView, ColorId.filling, this);
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsBlockedController.java:224]: Does this change address some issue or just removes duplicate filling draw?

File: app/src/main/java/org/thunderdog/challegram/telegram/TdlibCache.java [hunk 8]
```diff
@@ -1898,6 +1913,7 @@ private int putSupergroup (TdApi.Supergroup supergroup) {
 
   private boolean putSupergroupFull (long supergroupId, TdApi.SupergroupFullInfo supergroupFull) {
     supergroupsFulls.put(supergroupId, supergroupFull);
+    supergroupsFullsLastUpdateTime.append(supergroupId, SystemClock.uptimeMillis());
```
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/telegram/TdlibCache.java:1916]: Incorrect method used.

```suggestion
    supergroupsFullsLastUpdateTime.put(supergroupId, SystemClock.uptimeMillis());
```

**Review by vkryl — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java
Comment (vkryl) [inline app/src/main/java/org/thunderdog/challegram/ui/SettingsPrivacyKeyController.java:1]: Unfinished fix.

**Review by vkryl — CHANGES_REQUESTED** [review]

**Review by vkryl — APPROVED** [review]

**Review by christceder — COMMENTED** [review]
🙏🏻👍🏻

**Comment by Maybe778** [comment]
> - Updated design of restricted and unsupported messages
> - Slow Mode support
> - Media-only chats support
> - Improved interface for rights setting in groups
> - Public Photo support
> - Picker for avatars

**Comment by Maybe778** [comment]
> - Updated design of restricted and unsupported messages
> - Slow Mode support
> - Media-only chats support
> - Improved interface for rights setting in groups
> - Public Photo support
> - Picker for avatars

File: app/src/main/java/org/thunderdog/challegram/component/dialogs/ChatView.java
Comment (Maybe778) [reply app/src/main/java/org/thunderdog/challegram/component/dialogs/ChatView.java:1]: Maybe778

**Pull request merged**